### ISSUE TITLE
Repo is growing and growing

### ISSUE THREAD

**Issue by mitar** [ipfs/kubo#3789]
#### Version information:

```
go-ipfs version: 0.4.5-
Repo version: 5
System version: amd64/darwin
Golang version: go1.7.4
```

#### Description:

When creating a hashchain of objects (so each object links to a previous object), it seems repository gets much larger than expected. And it also just continues to grow and grow.

Reproduction: https://github.com/mitar/ipfs-create-chain

I started with empty repository (not sure why it is not completely empty, because I initialized it with `-e`):

```
$ date ; ./ipfs repo stat
Wed Mar 15 03:00:10 PDT 2017
NumObjects 	 5
RepoSize 	 334034
RepoPath 	 /Users/user/.ipfs
Version 	 fs-repo@5
```

After adding 10000 objects of a hashchain:

```
$ date ; ./ipfs repo stat
Wed Mar 15 03:01:16 PDT 2017
NumObjects 	 10005
RepoSize 	 2142370
RepoPath 	 /Users/user/.ipfs
Version 	 fs-repo@5
```

The last object in this hashchain is saying that the whole size is 1008672. 2142370 is more than twice as much.

After pinning the hashchain size jumps:

```
$ date ; ./ipfs repo stat
Wed Mar 15 03:01:24 PDT 2017
NumObjects 	 10267
RepoSize 	 5637553
RepoPath 	 /Users/user/.ipfs
Version 	 fs-repo@5
```

But leaving it running for 10 minutes just grows the repository:

```
$ date ; ./ipfs repo stat
Wed Mar 15 03:13:36 PDT 2017
NumObjects 	 10267
RepoSize 	 7167765
RepoPath 	 /Users/user/.ipfs
Version 	 fs-repo@5
```

After few more minutes:

```
$ date ; ./ipfs repo stat
Wed Mar 15 03:35:51 PDT 2017
NumObjects 	 10267
RepoSize 	 7452579
RepoPath 	 /Users/user/.ipfs
Version 	 fs-repo@5
```

And after few more hours:

```
$ date ; ./ipfs repo stat
Wed Mar 15 11:46:51 PDT 2017
NumObjects 	 10267
RepoSize 	 18612219
RepoPath 	 /Users/user/.ipfs
Version 	 fs-repo@5
```

What is happening here? This is a lot more than 1008672. OK, the list of all pins is probably around 1 MB more. So I would expect the repo to be around 2 MB.

There was no other use of the IPFS & daemon after the run of the initial script. I just left it running.

**Comment by mitar** [comment]
I also tested in on 0.4.6 and it looks the same.

**Comment by Kubuxu** [comment]
We are probably doing pins incrementally even if they are provided as one they will be pinned one by one. This means that there will be a lot of left over pin blocks, can you try running GC?

**Comment by mitar** [comment]
But it is also growing when I am not doing anything anymore.

GC removed 4 objects, it made it a bit smaller, but it is continuing growing.

**Comment by mitar** [comment]
Also, in my real project I am pinning one by one through time. So when a block is mined I add it to objects. And if it wins, I pin it (all peers pin the winning block).

**Comment by jefft0** [comment]
Can you check if it is the blocks or the datastore that is growing while the daemon is "just running"?

    du -d 0 /Users/user/.ipfs/blocks
    du -d 0 /Users/user/.ipfs/datastore

**Comment by mitar** [comment]
It is datastore growing:

```
$ du -d 0 blocks
43776	blocks
$ du -d 0 datastore/
1496	datastore/
```

After few minutes:

```
$ du -d 0 blocks
43776	blocks
$ du -d 0 datastore
1504	datastore
```

**Comment by Kubuxu** [comment]
The datastore directory is expected to grow as it contains cached providers and peer records. 
Those are not cleaned up as of now. On my node that is almost a year old they take only about 30MB.

**Comment by whyrusleeping** [comment]
My bootstrapper node mars has 110MB in its datastore. Its storing dht information, provider records, and peer address info. This gets gabrage collected periodically so it doesnt grow unboundedly.

If we fix the ephemeral address issue this will get a good deal smaller

**Comment by mitar** [comment]
So in production, where I run `ipfs daemon` with `--enable-gc`, my stats are after a week or so of running it:

```
$ ipfs repo stat
NumObjects 	 255220
RepoSize 	 1680016676
RepoPath 	 /ipfs
Version 	 fs-repo@5
```
```
$ ipfs pin ls | wc -l 
27481
```

As you see, there is many many more objects (255220) than they are pinned (27481).

Size of all pinned objects:

```
$ ipfs pin ls -q | xargs -n 1 ipfs object stat | grep BlockSize | awk '{s+=$2} END {print s}'
10702297
```

Again, as you see, there is huge difference between 10702297 and 1680016676.

**Comment by whyrusleeping** [comment]
@mitar does manually running a gc clear up the excess objects? If so, then its likely that the automatic garbage collection code is hitting a snag for some reason. We should look into that

**Comment by mitar** [comment]
How often should the automatic gc run?

**Comment by whyrusleeping** [comment]
Its set in the config file, by default its once per hour, and only runs if your repo is over 10GB

**Comment by mitar** [comment]
Aha. My repo is not yet over 10 GB.

**Comment by mitar** [comment]
Before `gc`:

```
$ ipfs repo stat
NumObjects 	 256308
RepoSize 	 1683546207
RepoPath 	 /ipfs
Version 	 fs-repo@5
```

After:

```
$ ipfs repo stat
NumObjects 	 27794
RepoSize 	 122828928
RepoPath 	 /ipfs
Version 	 fs-repo@5
```

This is much better. Still 122,828,928 >> 10,702,297.

And still:

```
du -sh /ipfs/
219M	/ipfs
```

Real size is larger than what `repo stat` returns.

**Comment by mitar** [comment]
One more example, after some time.

```
$ ipfs repo stat
NumObjects       592077
RepoSize         5447712568
RepoPath         /ipfs
Version          fs-repo@5
```

```
$ ipfs pin ls | wc -l
301392
```

```
$ ipfs pin ls -q | xargs -n 1 ipfs object stat | grep BlockSize | awk '{s+=$2} END {print s}'
117563843
```

117,563,843 vs 5,447,712,568. ;-)

**Comment by whyrusleeping** [comment]
@mitar does running an ipfs repo gc help?

**Comment by mitar** [comment]
Before:

```
NumObjects 	 627413
RepoSize 	 6079495728
RepoPath 	 /ipfs
Version 	 fs-repo@5
```

After:

```
NumObjects 	 301660
RepoSize 	 340919726
RepoPath 	 /ipfs
Version 	 fs-repo@5
```

After gc is 340,919,726, much better, but still 3x. But I am OK with this.

So this is still just an issue of automatic gc not running until 10 GB. OK.

**Comment by whyrusleeping** [comment]
@mitar yeah, you could set up a cron job to run a gc every so often until we have better gc automation

**Comment by antonzhukov** [comment]
The same problem I got. Even if when I run ipfs with gc enabled and 10Gb max storage size, no gc actually happens.

I am running ipfs in docker:

```
docker-compose exec ipfs ipfs config show | grep -E "GC|Storage"
    "GCPeriod": "1h",
    "StorageGCWatermark": 90,
    "StorageMax": "10GB"
```

```
docker-compose exec ipfs ps aux | grep ipfs
    1 root      0:06 /sbin/tini -- /usr/local/bin/start_ipfs daemon --enable-gc
    9 ipfs      1d11 ipfs daemon --enable-gc --migrate=true
```
```
du -sh .ipfsdata/
85G	.ipfsdata/
```

**Comment by Stebalien** [comment]
Have you added files to your node? GC won't remove pinned data.

**Comment by mitar** [comment]
See above for reproduction. This is more than just a question about GC and pining.

**Comment by Stebalien** [comment]
@mitar your issue is but 90GiB is a bit excessive (unless it's just that GC hasn't run yet).

**Comment by mitar** [comment]
@Stebalien I do not understand what you are saying. The issue here is that after I initially populate IPFS with data and leave IPFS running the size it consumes is growing and growing. Even without GC this should not be happening, no?
